### DISCUSSION TITLE
Suspicion of bug in the sorting of helmfile values fields

### DISCUSSION THREAD

**Discussion by bgiegel** [helmfile/helmfile#1547] [Questions]
Hello ! 

I’m trying to debug a weird issue I have on one of my helmfile config. Here is my setup :

backup.env.yaml:
```
opensearch:
  backups:
     "465": { client_id: "07c9725629384b6b", enabled: true, region: "gva" }
      "456": { client_id: "4553673c6164c419", enabled: true, region: "gva" }
      "433": { client_id: "4fe2ed584743bf9e", enabled: true, region: "gva" }

```
Here I have a map with an id as the key (4xx) and a client_id which is alphnumerical value.

backupd.values.yaml.gotmpl
```
config:
  backends:
    {{ range $tenantID, $backup :=  $opensearch.backups }}
    {{ $backup.client_id }}:
.... some other redacted fields here
    {{ end }}
``` 
Here in my values go template file I am ranging through the opensearch.backups map you can see in backup.env.yaml.
in the end it generates something like this when I run helmfile template : 

```
  4553673c6164c419:
  ...
  4fe2ed584743bf9e:
  ...
  07c9725629384b6b:
  ...
```

But the thing that drives me nuts is that the order is changing from one `helmfile template` to the other. The output is not deterministic... Sometimes it’s the 07c client id first sometimes it’s the 4fe.. one. I don’t get it. If I replace the client ids purely with alphabetical characters like abc,def,ghi in my backupd.env.yaml it works fine the output is always the same, same if I use any numbers. It’s really with some specific sequence of string that I have the issue... 

NOTE : I also have another file with default values where I add static values to the backends map (just so that you know that there is a merge of these 2 at some point).

So it looks to me like a bug... I don’t know if it’s on the helmfile side or if it’s even lower but I can’t figure out what’s wrong in my code.

I tried to use `sortAlpha` function from the `Sprig` Library, but it’s the same. Something after the merge of all the values.yaml is sorting the fields by pseudo alphabetical order but the thing that sort this is buggy, or at least it look like it is...

This is very annoying because it means that helmfile is redeploying my release on each helmfile apply even though nothing changed...

Any idea ?

**Comment by yxxhero** [comment]
@bgiegel why not use array format:
```
backips:
  - fsadfas
  - fsdfa
  - fsadfa
```

**Comment by yxxhero** [comment]
and pleases post an full example.

**Comment by bgiegel** [comment]
Thank you for your answer. 

I didn’t used an array in the input values because a map seem more natural and closer to the output config. The output config is a configmap. It’s a config file of our application and they use maps. I can change the format off course but it’s more work for me and also it’s not the only app to use that format. And the input "backups" field is not what causing the issue. It’s really the ouput "backends" map that is not sorted correctly.

The repo and the chart that I am using are private. But I’ll try to reproduce the example on a small public project.

**Comment by bgiegel** [comment]
https://github.com/bgiegel/helmfile-map-sorting-problem

Here I wrote a minimal helmfile config that reproduce the issue. You can just run the helmfile template command and see the output. 

Note that I have the same issue if I run helmfile write-values. Which means the problem is not due to the base chart, it’s really the generation of the merged values file that cause the issue.

**Reply by yxxhero** [reply]
@bgiegel yeah. I will see that and checkout what happend.

**Reply by yxxhero** [reply]
@bgiegel could you use helm template and see what will happend?

**Reply by yxxhero** [reply]
@bgiegel I think that it's a issue of helm. you can use helm to test this. and pass two values file.

**Reply by bgiegel** [reply]
I’ve tested with helm. I’ve put the content of the env file in values.yaml and the template with the range in the configmap manifest template directly and ran a `helm template backupd chart/backupd` and it always sort it in the same order.

output is always :  

```
      4fe2ed584743bf9e:
   ...   
      4553673c6164c419:
  ...    
      07c9725629384b6b:
  ...
```

No matter the order of the values.yaml map and how many helm template command I run.

Not that I also have the issue after a write-values and if I’m not mistaken this command doesn’t call an helm command right ?

**Reply by bgiegel** [reply]
I went through the helmfile code in debug mode and found that in the write-values command we have the "right" order (the same order as helm with 4fe2 always first) here : https://github.com/helmfile/helmfile/blob/0349197b20b8739ae384aaca263285f48eb2cd6e/pkg/state/state.go#L1609

when we unmarshall the values computed by the values that contains the `range`.

Them we merge it using this line : https://github.com/helmfile/helmfile/blob/0349197b20b8739ae384aaca263285f48eb2cd6e/pkg/state/state.go#L1613

and then we re-encode the merged file into the output file. Which end up sorting the map entries in a random order. So I suppose the issue comes from the yaml encoder... https://github.com/helmfile/helmfile/blob/0349197b20b8739ae384aaca263285f48eb2cd6e/pkg/state/state.go#L1621

**Reply by bgiegel** [reply]
Looks like you are using yaml v2. I tried switching to yaml v3 encoder and it seems to fix the issue : the output stay the same no matter how many times I run the helmfile command.

**Reply by yxxhero** [reply]
PR is welcome.

**Reply by bgiegel** [reply]
I’ll do that later today or tomorrow thanks.
